README.md
# @grandom/fnv1a

[Fowler–Noll–Vo][url-wikipedia] ***(FNV-1a)*** non-cryptographic hash function implementation written in TypeScript.

## Install

```
npm i @grandom/fnv1a
```

## License

[MIT][url-license-doc] @ [Richard King](https://richrdkng.com)

<!--- References =============================================================================== -->

<!--- URLs -->
[url-wikipedia]: https://en.wikipedia.org/wiki/Fowler-Noll-Vo_hash_function
[url-license-doc]: https://github.com/grandom-library/fnv1a/blob/main/LICENSE


CHANGELOG.md
## [1.0.7](https://github.com/grandom-library/fnv1a/compare/v1.0.6...v1.0.7) (2023-09-25)


### Performance Improvements

* **deps:** bump @types/jest from 29.5.4 to 29.5.5 ([424f589](https://github.com/grandom-library/fnv1a/commit/424f589da238cb3cfd3c659cc746d11180757b3d))

## [1.0.6](https://github.com/grandom-library/fnv1a/compare/v1.0.5...v1.0.6) (2023-09-14)


### Performance Improvements

* **deps:** bump jest from 29.6.4 to 29.7.0 ([7188d1c](https://github.com/grandom-library/fnv1a/commit/7188d1c233290e8d6bef696a0f531e1171923c5d))

## [1.0.5](https://github.com/grandom-library/fnv1a/compare/v1.0.4...v1.0.5) (2023-09-14)


### Performance Improvements

* **deps:** bump rollup from 3.29.0 to 3.29.1 ([2230f4b](https://github.com/grandom-library/fnv1a/commit/2230f4b6f896b6cda43e2b0b96f9c430ca43bf1f))

## [1.0.4](https://github.com/grandom-library/fnv1a/compare/v1.0.3...v1.0.4) (2023-09-07)


### Performance Improvements

* **deps:** bump rollup from 3.28.1 to 3.29.0 ([b530d8d](https://github.com/grandom-library/fnv1a/commit/b530d8d2a3c1e728b834993c9c40652e91c8b5f1))

## [1.0.3](https://github.com/grandom-library/fnv1a/compare/v1.0.2...v1.0.3) (2023-08-28)


### Performance Improvements

* **deps:** bump @rollup/plugin-typescript from 11.1.2 to 11.1.3 ([ab77e52](https://github.com/grandom-library/fnv1a/commit/ab77e524c3a03b35b178d6cc7277302ffd024a96))

## [1.0.2](https://github.com/grandom-library/fnv1a/compare/v1.0.1...v1.0.2) (2023-08-24)


### Performance Improvements

* **deps:** update dependencies ([d1b3069](https://github.com/grandom-library/fnv1a/commit/d1b306902277f6c4ef0ad304ae1bfd7bbaa5aaf6))

## [1.0.1](https://github.com/grandom-library/fnv1a/compare/v1.0.0...v1.0.1) (2023-08-24)


### Performance Improvements

* **deps:** bump jest from 29.6.3 to 29.6.4 ([9fe219b](https://github.com/grandom-library/fnv1a/commit/9fe219bd7e854b01a65c0ab2db5a89026b3eb397))

# 1.0.0 (2023-08-24)


### Features

* initial release ([b62d3f8](https://github.com/grandom-library/fnv1a/commit/b62d3f8e05c928c052caa2d6e5fd87ba1ce3dd6f))


test/index.test.ts
import fnv1a from '../src'

describe('errors', () => {
  test('1st arg must be a string', () => {
    // @ts-expect-error
    expect(() => fnv1a()).toThrowWithMessage(
      TypeError,
      'The \'string\' argument must be a string, got: undefined (typeof === \'undefined\').'
    )
  })

  test('2nd arg must be a number', () => {
    // @ts-expect-error
    expect(() => fnv1a('', false)).toThrowWithMessage(
      TypeError,
      'The \'hashSize\' argument must be a number, got: false (typeof === \'boolean\').'
    )
  })

  test('2nd arg must be one of 32, 64, 128, 256, 512, 1024', () => {
    expect(() => fnv1a('', 48)).toThrowWithMessage(
      RangeError,
      'The \'hashSize\' argument must be one of 32, 64, 128, 256, 512, 1024, got: 48.'
    )
  })
})

describe('number', () => {
  test('number - 32 bit', () => {
    expect(fnv1a('')).toBe(2166136261)
    expect(fnv1a('a')).toBe(3826002220)
    expect(fnv1a('Iscx9cE2rk1YMvPzpb4NFCUiLOiXmm2l')).toBe(2773037204)
  })
})

describe('bigint', () => {
  test('bigint - 32 bit', () => {
    expect(fnv1a('', 32).toString()).toBe(2166136261n.toString())
    expect(fnv1a('a', 32).toString()).toBe(3826002220n.toString())
    expect(fnv1a('Iscx9cE2rk1YMvPzpb4NFCUiLOiXmm2l', 32).toString()).toBe(2773037204n.toString())
  })

  test('bigint - 64 bit', () => {
    expect(fnv1a('', 64).toString()).toBe(14695981039346656037n.toString())
    expect(fnv1a('a', 64).toString()).toBe(12638187200555641996n.toString())
  })

  test('bigint - 128 bit', () => {
    expect(fnv1a('', 128).toString()).toBe(144066263297769815596495629667062367629n.toString())
    expect(fnv1a('a', 128).toString()).toBe(279349696671665264231447081680185756004n.toString())
  })

  test('bigint - 256 bit', () => {
    expect(fnv1a('', 256).toString()).toBe(100029257958052580907070968620625704837092796014241193945225284501741471925557n.toString())
    expect(fnv1a('a', 256).toString()).toBe(44867753943577790287705335483255842241845894581843275463348238411130865804156n.toString())
  })

  test('bigint - 512 bit', () => {
    expect(fnv1a('', 512).toString()).toBe(9659303129496669498009435400716310466090418745672637896108374329434462657994582932197716438449813051892206539805784495328239340083876191928701583869517785n.toString())
    expect(fnv1a('a', 512).toString()).toBe(11953317394135539349154433287628773929568787445269509959212693531458072457957106370775899041675541513348796391014922741597244887016216087954051796457488264n.toString())
  })

  test('bigint - 1024 bit', () => {
    expect(fnv1a('', 1024).toString()).toBe(14197795064947621068722070641403218320880622795441933960878474914617582723252296732303717722150864096521202355549365628174669108571814760471015076148029755969804077320157692458563003215304957150157403644460363550505412711285966361610267868082893823963790439336411086884584107735010676915n.toString())
    expect(fnv1a('a', 1024).toString()).toBe(5818363527381644214167233747835538701107978818630109145693299645251540493578435218053541778827587683654772935121139775523503145629158537612850888831472542273057513046212368429512907352387112346131801901403545957510684065167101045719835285505190875790984075398467686988232751512038527112618n.toString())
  })
})


src/index.ts
const _32_BIT_OFFSET = 2166136261

const PRIMES = {
  32: BigInt(2 ** 24 + 2 ** 8 + 0x93),
  64: BigInt(2 ** 40 + 2 ** 8 + 0xb3),
  128: BigInt(2) ** BigInt(88) + BigInt(2 ** 8 + 0x3b),
  256: BigInt(2) ** BigInt(168) + BigInt(2 ** 8 + 0x63),
  512: BigInt(2) ** BigInt(344) + BigInt(2 ** 8 + 0x57),
  1024: BigInt(2) ** BigInt(680) + BigInt(2 ** 8 + 0x8d)
} as const

const OFFSETS: Record<keyof typeof PRIMES, bigint> = {
  32: BigInt(_32_BIT_OFFSET),
  64: 14695981039346656037n,
  128: 144066263297769815596495629667062367629n,
  256: 100029257958052580907070968620625704837092796014241193945225284501741471925557n,
  512: 9659303129496669498009435400716310466090418745672637896108374329434462657994582932197716438449813051892206539805784495328239340083876191928701583869517785n,
  1024: 14197795064947621068722070641403218320880622795441933960878474914617582723252296732303717722150864096521202355549365628174669108571814760471015076148029755969804077320157692458563003215304957150157403644460363550505412711285966361610267868082893823963790439336411086884584107735010676915n
} as const

/**
 * Returns a 32-bit FNV-1a hash.
 *
 * @param string The string to hash.
 */
function fnv1a (string: string): number

/**
 * Returns an n-bit FNV-1a hash.
 *
 * @param string The string to hash.
 * @param hashSize The size of the hash.
 */
function fnv1a (string: string, hashSize: number): bigint

function fnv1a (arg1: any, arg2?: any): any {
  if (typeof arg1 !== 'string') {
    throw new TypeError(
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      `The 'string' argument must be a string, got: ${arg1} (typeof === '${typeof arg1}').`
    )
  }

  const length = arg1.length

  // 32 bit hash as number -------------------------------------------------------------------------
  if (typeof arg2 === 'undefined') {
    let hash = _32_BIT_OFFSET

    for (let i = 0; i < length; i++) {
      hash ^= arg1.charCodeAt(i)
      hash += (hash << 1) + (hash << 4) + (hash << 7) + (hash << 8) + (hash << 24)
    }

    return hash >>> 0
  }

  if (typeof arg2 !== 'number') {
    throw new TypeError(
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      `The 'hashSize' argument must be a number, got: ${arg2} (typeof === '${typeof arg2}').`
    )
  }

  if (!(arg2 in PRIMES)) {
    throw new RangeError(
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      `The 'hashSize' argument must be one of 32, 64, 128, 256, 512, 1024, got: ${arg2}.`
    )
  }

  // n-bit hash as BigInt --------------------------------------------------------------------------
  const size = arg2 as keyof typeof PRIMES
  const prime = PRIMES[size]

  let hash = OFFSETS[size]

  for (let i = 0; i < length; i++) {
    hash ^= BigInt(arg1.charCodeAt(i))
    hash = BigInt.asUintN(size, hash * prime)
  }

  return hash
}

export default fnv1a
